### PR TITLE
fix(router): should reset location if a navigation by location is suc…

### LINKED ISSUE #13491: First routing to module fails when user navigates by hash changing
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
After upgrading from 2.1.0-beta.0 to 2.3.1, we are faced with the following problem: router navigates to previous page when user follow to new page by clicking a-link. It's reproduced once for pages which presents as modules. When module is loaded on browser everything is fine. It's OK for angular versions 2.1.0-beta.0 and less.
Our specific case is navigation by a-link because we have portal application which integrated to another complex web-application. And navigation is usually initiated by not-angular app within clicking menu links.

**Expected behavior**
Any navigation initiated by address changes should be completed successfully.

**Minimal reproduction of the problem with instructions**
Plunker: https://plnkr.co/edit/hpkyfVyc3WLdWortxXFT
1. Click button "Page 1" - page 1 is loaded, URL is correct.
2. Click link "Page 2" - page 2 is loaded, but URL is wrong.

**What is the motivation / use case for changing the behavior?**
NA

**Please tell us about your environment:** 
Ubuntu 14.04, NPM, Webpack, Node 7.x 

* **Angular version:** 
2.3.1

* **Browser:** 
all 
 
* **Language:** 
all

### PR DESCRIPTION
**Pull request by vicb**
Supersedes #13545 by also fixing the `PathLocationStrategy`.

/cc @vsavkin 
@chuckjaz please do NOT use GH to merge this one (for attribution)

Closes #13491
Closes #13545
Closes #13574

### CODE REVIEW

**Comment by DzmitryShylovich** [comment]
I believe it also fixes https://github.com/angular/angular/issues/12438

File: modules/@angular/common/src/location/hash_location_strategy.ts [hunk 1]
```diff
@@ -45,10 +45,7 @@ export class HashLocationStrategy extends LocationStrategy {
     }
   }
 
-  onPopState(fn: LocationChangeListener): void {
-    this._platformLocation.onPopState(fn);
-    this._platformLocation.onHashChange(fn);
-  }
+  onPopState(fn: LocationChangeListener): void { this._platformLocation.onPopState(fn); }
```
Comment (DzmitryShylovich) [inline modules/@angular/common/src/location/hash_location_strategy.ts]: I think it should be `this._platformLocation.onHashChange(fn);` in `HashLocationStarategy` and `this._platformLocation.onPopState(fn);` in `PathLocationStrategy`

Comment (vicb) [inline modules/@angular/common/src/location/hash_location_strategy.ts]: @DzmitryShylovich what are the reasons you think so ?

Comment (DzmitryShylovich) [inline modules/@angular/common/src/location/hash_location_strategy.ts]: 1. Readability `HashLocationStarategy` - `onHashChange`.
2. PopState is a part of History api, so I think `HashChange` will probably has a better support across old/mobile browsers (there're some existing issues related to mobile ie and history api, for example).

**Comment by vicb** [comment]
@vsavkin came with a better fix in #13545

**Pull request closed without merging**